File: python/1-50/47. Permutations II.py
class Solution(object):
    def permuteUnique(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        if not nums:
            return []
        ret = [[]]
        for n in nums:
            new_ret = []
            l = len(ret[0])
            for seq in ret:
                for i in range(l, -1, -1):
                    if i < l and seq[i] == n:
                        break
                    new_ret.append(seq[:i] + [n] + seq[i:])
                ret = new_ret
            print new_ret
        return ret

File: cpp/1-50/47. Permutations II.cpp
class Solution {
public:
	vector<vector<int>> permuteUnique(vector<int>& nums)
	{
		vector<vector<int>> res;
		sort(nums.begin(), nums.end());
		dfs(0, nums, res);
		return res;
	}
	void dfs(int pos, vector<int> &nums, vector<vector<int>> &res)
	{
		if (pos == nums.size() - 1) {
			res.push_back(nums);
			return;
		}
		for (int i = pos; i < nums.size(); i++)
		{
			if (i != pos && nums[i] == nums[pos])
				continue;
			swap(nums[i], nums[pos]);
			dfs(pos + 1, nums, res);
			swap(nums[i], nums[pos]);
		}
	}
};